// zcatalyst-cli/lib/endpoints/lib/tunnel.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const error_1 = __importDefault(require("../../error"));
const api_1 = __importDefault(require("../../internal/api"));
const logger_1 = require("../../util_modules/logger");
class TunnelAPI {
    constructor(projectId, opts) {
        this.projectId = projectId;
        this.opts = opts;
    }
    getTunnel() {
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield new api_1.default(this.opts).get(`/baas/v1/project/${this.projectId}/tunneling`);
            if (res.body && res.body.data) {
                return res.body.data;
            }
            (0, logger_1.debug)('get tunnel response from server : ' + res.body);
            throw new error_1.default('Server Error: Unexpected Response from server.', {
                exit: 2
            });
        });
    }
    enableTunnel(tunnelUrl) {
        return __awaiter(this, void 0, void 0, function* () {
            const qs = {};
            if (tunnelUrl) {
                qs.tunnel_url = tunnelUrl;
            }
            const res = yield new api_1.default(this.opts).patch(`/baas/v1/project/${this.projectId}/tunneling`, {
                qs
            });
            if (res.body && res.body.data) {
                return res.body.data;
            }
            (0, logger_1.debug)('enable tunnel response from server : ' + res.body);
            throw new error_1.default('Server Error: Unexpected Response from server.', {
                exit: 2
            });
        });
    }
    disableTunnel() {
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield new api_1.default(this.opts).patch(`/baas/v1/project/${this.projectId}/tunneling`);
            if (res.body && res.body.data) {
                return res.body.data;
            }
            (0, logger_1.debug)('disable tunnel response from server : ' + res.body);
            throw new error_1.default('Server Error: Unexpected Response from server.', {
                exit: 2
            });
        });
    }
    tunnelCallback(status, details, data) {
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield new api_1.default(this.opts).post(`/baas/v1/project/${this.projectId}/tunneling/callback`, {
                body: {
                    status,
                    content: Object.assign({}, details),
                    data
                }
            });
            if (res.status >= 200 && res.status <= 299) {
                return;
            }
            (0, logger_1.debug)(`tunnel callback response from server ::: status=${res.status} ::: body=${res.body}`);
            throw new error_1.default('Server Error: Unexpected Response from server.', {
                exit: 2
            });
        });
    }
}
exports.default = TunnelAPI;
