// zcatalyst-sdk-node/lib/stratus/util.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _Util_sign;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Util = void 0;
const constants_1 = __importDefault(require("../utils/constants"));
const qs_1 = __importDefault(require("qs"));
const { REQ_METHOD, CREDENTIAL_USER } = constants_1.default;
class Util {
    constructor(bucket) {
        var _a, _b;
        _Util_sign.set(this, void 0);
        this.bucket = bucket;
        this._requester = bucket.getAuthorizationClient();
        this.userType = (_a = this._requester.app) === null || _a === void 0 ? void 0 : _a.credential.getCurrentUserType();
        this.userScope = (_b = this._requester.app) === null || _b === void 0 ? void 0 : _b.credential.getCurrentUser();
    }
    _getUserType() {
        return this.userType;
    }
    _getUserScope() {
        return this.userScope;
    }
    getBucketSignature() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.userType === 'admin' || this.userScope === 'admin') {
                if (__classPrivateFieldGet(this, _Util_sign, "f") && __classPrivateFieldGet(this, _Util_sign, "f").expiry_time > Date.now()) {
                    return qs_1.default.parse(__classPrivateFieldGet(this, _Util_sign, "f").signature);
                }
                const request = {
                    method: REQ_METHOD.post,
                    path: '/bucket/signature',
                    qs: { bucket_name: this.bucket.bucketDetails.bucket_name },
                    type: "json" /* RequestType.JSON */,
                    expecting: "json" /* ResponseType.JSON */,
                    catalyst: true,
                    track: true,
                    user: CREDENTIAL_USER.admin
                };
                const resp = (yield this._requester.send(request)).data.data;
                __classPrivateFieldSet(this, _Util_sign, resp, "f");
                return qs_1.default.parse(resp.signature);
            }
            return undefined;
        });
    }
}
exports.Util = Util;
_Util_sign = new WeakMap();
