// zcatalyst-cli/lib/serve/server/lib/master/unknown-req-proxy.js
'use strict';
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const runtime_store_1 = __importDefault(require("../../../../runtime-store"));
const logger_1 = require("../../../../util_modules/logger");
exports.default = (proxyInstance, masterPort, customProxyUrl) => (req, res) => {
    if (req.url.length === 1 && req.url.startsWith('/')) {
        res.redirect('/app/');
    }
    else if (req.url.match(/accounts\/p\/[0-9]+\/clientidprequest/g)) {
        res.redirect(308, `https://${req.headers['x-zc-project-domain']}${req.url}`);
        return;
    }
    else if (typeof customProxyUrl === 'string' &&
        !req.url.match(/^(\/baas|\/app|\/server|\/oauthorize|\/__catalyst).*/g)) {
        (0, logger_1.debug)(`proxying "${req.url}" to "${customProxyUrl}"`);
        proxyInstance.web(req, res, {
            target: customProxyUrl,
            ws: true,
            changeOrigin: true,
            secure: false,
            headers: {
                'User-Agent': runtime_store_1.default.get('context.cli.package.name') +
                    '/' +
                    runtime_store_1.default.get('context.cli.package.version')
            }
        });
    }
    else {
        const headers = {};
        if (req.url.match(/accounts\/p\/[0-9]+\/webclient\/v1\/captcha/g)) {
            headers['origin'] = `https://${req.headers['x-zc-project-domain']}`;
        }
        if (req.url.includes('/__catalyst/auth/')) {
            headers['catalyst-redirect-domain'] = `http://localhost:${masterPort || 3000}`;
        }
        proxyInstance.web(req, res, {
            target: `https://${req.headers['x-zc-project-domain']}`,
            changeOrigin: true,
            cookieDomainRewrite: {
                [req.headers['x-zc-project-domain'] + '']: 'localhost'
            },
            ws: true,
            headers
        });
    }
};
