// zcatalyst-sdk-node/lib/stratus/transfer-manager/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __await = (this && this.__await) || function (v) { return this instanceof __await ? (this.v = v, this) : new __await(v); }
var __asyncGenerator = (this && this.__asyncGenerator) || function (thisArg, _arguments, generator) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var g = generator.apply(thisArg, _arguments || []), i, q = [];
    return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
    function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
    function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
    function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
    function fulfill(value) { resume("next", value); }
    function reject(value) { resume("throw", value); }
    function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
};
var _TransferManager_instances, _TransferManager_uploadPart, _TransferManager_promisePool, _TransferManager_getObjectPart;
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransferManager = void 0;
const console_1 = require("console");
const validator_1 = require("../../utils/validator");
const error_1 = require("../../utils/error");
const multipart_upload_1 = require("./multipart-upload");
class TransferManager {
    constructor(bucketInstance) {
        _TransferManager_instances.add(this);
        this._requester = bucketInstance.getAuthorizationClient();
        this.bucket = bucketInstance;
    }
    /**
     * Create a multipart instance.
     * @param objectName name of the object.
     * @returns MultipartUpload.
     */
    createMultipartInstance(objectName, uploadId) {
        return __awaiter(this, void 0, void 0, function* () {
            yield (0, validator_1.wrapValidatorsWithPromise)(() => {
                (0, validator_1.isNonEmptyString)(objectName, 'object_name', true);
            }, error_1.CatalystStratusError);
            if (uploadId === undefined) {
                uploadId = (yield this.bucket.initiateMultipartUpload(objectName)).upload_id;
            }
            return new multipart_upload_1.MultipartUpload(this.bucket, objectName, uploadId);
        });
    }
    /**
     * Upload an object as multiple parts.
     * @param data object body to be uploaded.
     * @param partSize size of the each object part.
     * @returns string.
     */
    putObjectAsParts(objectName, data, partSize, concurrency = 5) {
        return __awaiter(this, void 0, void 0, function* () {
            yield (0, validator_1.wrapValidatorsWithPromise)(() => {
                (0, validator_1.isNonEmptyString)(objectName, 'object_name', true);
                (0, validator_1.isNonNullValue)(data, 'object_body', true);
            }, error_1.CatalystStratusError);
            partSize = partSize * (1024 * 1024);
            const initiateRes = yield this.createMultipartInstance(objectName);
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            const parts = [];
            let partNumber = 1;
            return new Promise((resolve, reject) => {
                data.on('readable', () => {
                    let chunk;
                    while ((chunk = data.read(partSize)) !== null) {
                        ((part, partNum) => {
                            parts.push(() => __classPrivateFieldGet(this, _TransferManager_instances, "m", _TransferManager_uploadPart).call(this, initiateRes, part, partNum, partSize));
                        })(chunk, partNumber++);
                    }
                });
                data.on('end', () => __awaiter(this, void 0, void 0, function* () {
                    try {
                        yield __classPrivateFieldGet(this, _TransferManager_instances, "m", _TransferManager_promisePool).call(this, parts, concurrency);
                        const completeRes = yield initiateRes.completeUpload();
                        if (!completeRes) {
                            return reject(`Error when completing the multipart object(${objectName}) upload`);
                        }
                        (0, console_1.info)('Upload completed');
                        resolve('true');
                    }
                    catch (err) {
                        reject(err);
                    }
                }));
                data.on('error', (err) => {
                    reject(`Error reading file: ${err.message}`);
                });
            });
        });
    }
    /**
     * Get the object as multiple parts
     * @param partSize in MB. Size of each parts.
     * @returns AsyncGenerator<Buffer, void>.
     */
    getIterableObject(objectName, partSize) {
        return __asyncGenerator(this, arguments, function* getIterableObject_1() {
            yield __await((0, validator_1.wrapValidatorsWithPromise)(() => {
                (0, validator_1.isNonEmptyString)(objectName, 'object_name', true);
            }, error_1.CatalystStratusError));
            partSize = partSize * 1024 * 1024;
            const file = yield __await(this.bucket.object(objectName).getDetails());
            const fileSize = file.size;
            let start = 0;
            let end = partSize;
            while (start < fileSize) {
                const chunk = yield __await(__classPrivateFieldGet(this, _TransferManager_instances, "m", _TransferManager_getObjectPart).call(this, objectName, start, end));
                const chunks = [];
                yield yield __await(new Promise((resolve, reject) => {
                    chunk.on('data', (data) => {
                        chunks.push(data);
                    });
                    chunk.on('end', () => resolve(Buffer.concat(chunks)));
                    chunk.on('error', reject);
                }));
                end = (Math.min(start + partSize, fileSize) - 1);
                start = end + 1;
            }
        });
    }
    /**
     * Get the object as multiple parts.
     * @param partSize in MB. Size of each parts.
     * @returns downloadable object parts.
     */
    generatePartDownloaders(objectName, partSize) {
        return __awaiter(this, void 0, void 0, function* () {
            yield (0, validator_1.wrapValidatorsWithPromise)(() => {
                (0, validator_1.isNonEmptyString)(objectName, 'object_name', true);
            }, error_1.CatalystStratusError);
            const parts = [];
            partSize = partSize * 1024 * 1024;
            const file = yield this.bucket.object(objectName).getDetails();
            const fileSize = file.size;
            let start = 0;
            while (start < fileSize) {
                const end = (Math.min(start + partSize, fileSize) - 1);
                ((startVal) => {
                    parts.push(() => __classPrivateFieldGet(this, _TransferManager_instances, "m", _TransferManager_getObjectPart).call(this, objectName, startVal, end));
                })(start);
                start = end + 1;
            }
            return parts;
        });
    }
}
exports.TransferManager = TransferManager;
_TransferManager_instances = new WeakSet(), _TransferManager_uploadPart = function _TransferManager_uploadPart(initiateRes, stream, partNumber, partSize) {
    return __awaiter(this, void 0, void 0, function* () {
        const res = yield initiateRes.uploadPart(stream, partNumber);
        if (res) {
            (0, console_1.info)(`Part ${partNumber} Uploaded`);
        }
        else {
            throw (0, console_1.error)(`Error while uploading the part ${(partNumber - 1) * partSize} to ${partNumber * partSize} in bytes`);
        }
        return res;
    });
}, _TransferManager_promisePool = function _TransferManager_promisePool(tasks, concurrency) {
    return __awaiter(this, void 0, void 0, function* () {
        const executing = [];
        for (const task of tasks) {
            const p = task().then(() => {
                executing.splice(executing.indexOf(p), 1);
            });
            executing.push(p);
            if (executing.length >= concurrency) {
                yield Promise.race(executing);
            }
        }
        return yield Promise.all(executing);
    });
}, _TransferManager_getObjectPart = function _TransferManager_getObjectPart(objectName, start, end, retries = 3) {
    return __awaiter(this, void 0, void 0, function* () {
        yield (0, validator_1.wrapValidatorsWithPromise)(() => {
            (0, validator_1.isNonEmptyString)(objectName, 'object_name', true);
        }, error_1.CatalystStratusError);
        const res = yield this.bucket.getObject(objectName, { range: `${start}-${end}` });
        const err = new Promise((reject) => {
            res.on('error', () => __awaiter(this, void 0, void 0, function* () {
                if (retries > 0) {
                    yield new Promise((resolve) => setTimeout(resolve, 1000));
                    yield __classPrivateFieldGet(this, _TransferManager_instances, "m", _TransferManager_getObjectPart).call(this, objectName, start, end, --retries);
                }
                else {
                    reject(`Error when downloading the part from ${start}-${end} bytes`);
                }
            }));
        });
        return res || err;
    });
};
