// zcatalyst-cli/lib/util_modules/constants/lib/default.js
'use strict';
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const dc_type_1 = __importDefault(require("./dc-type"));
exports.default = Object.freeze({
    dc: dc_type_1.default.us.value,
    env_name: 'Development',
    catalystError: 'CatalystError',
    scripts: ['predeploy', 'postdeploy', 'preserve', 'postserve'],
    serve_port: {
        http: {
            master: 3000,
            advancedio: 4000,
            basicio: 4200,
            browser_logic: 4400,
            appsail: {
                master: 3000,
                service: 4600
            },
            client: 5000
        },
        debug: {
            master: -1,
            advancedio: 8000,
            basicio: 8200,
            browser_logic: 8400,
            appsail: -1,
            client: -1
        }
    },
    event_bus: {
        name: 'Default'
    },
    datastore: {
        column_details: [
            {
                column_name: 'ROWID',
                data_type: 'bigint'
            },
            {
                column_name: 'CREATORID',
                data_type: 'bigint'
            },
            {
                column_name: 'CREATEDTIME',
                data_type: 'datetime'
            },
            {
                column_name: 'MODIFIEDTIME',
                data_type: 'datetime'
            }
        ]
    }
});
