// zcatalyst-cli/lib/event_generate/stratus.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const moment_1 = __importDefault(require("moment"));
const runtime_store_1 = __importDefault(require("../runtime-store"));
const constants_1 = require("../util_modules/constants");
const js_1 = require("../util_modules/js");
const project_1 = require("../util_modules/project");
function generateStratusEventJson(eventJson) {
    const userDetails = {
        email_id: runtime_store_1.default.get('user.Email'),
        user_id: js_1.JS.randomNumber(15),
        is_confirmed: getRandomBoolean(),
        last_name: js_1.JS.randomString(5),
        first_name: js_1.JS.randomString(10),
        zuid: js_1.JS.randomNumber(10)
    };
    const time = (0, moment_1.default)().format('MMM DD, YYYY hh:mm A');
    const data = {
        bucket_details: {
            bucket_name: 'sample-bucket',
            project_details: {
                id: (0, project_1.getProjectId)(),
                project_name: (0, project_1.getProjectName)(),
                project_type: 'Live'
            },
            created_by: userDetails,
            created_time: time,
            modified_by: userDetails,
            modified_time: time,
            bucket_meta: {
                versioning: getRandomBoolean(),
                caching: {
                    status: getRandomBoolean() ? 'Enabled' : 'Disabled'
                },
                encryption: getRandomBoolean(),
                audit_consent: getRandomBoolean()
            }
        },
        object_key: 'path1/path2/myfile.png'
    };
    eventJson = eventJson.replace(new RegExp(constants_1.PLACEHOLDER.event_data.rule_id, 'g'), 'NULL');
    eventJson = eventJson.replace(new RegExp(constants_1.PLACEHOLDER.event_data.source_id, 'g'), 'NULL');
    eventJson = eventJson.replace(new RegExp(constants_1.PLACEHOLDER.event_data.data, 'g'), JSON.stringify(data));
    return eventJson;
}
exports.default = generateStratusEventJson;
function getRandomBoolean() {
    return js_1.JS.random(0, 1) === 1;
}
